### DISCUSSION TITLE
How to add LineString to existing LineString layer dynamically

### DISCUSSION THREAD

**Discussion by gigfy2** [Mapsui/Mapsui#2555] [Q&A]
I can create a LineString layer with lines, but I don't see a way to add lines as time goes on to that same layer.  Every example I've seen adds all the lines when the layer is created.  Is there an example of adding lines to an existing LineString layer and not just when the layer is created?

**Comment by pauldendulk** [comment]
The idea is that Mapsui should not become a data access layer. That part should be under the users control. Mapsui is a way to visualize that data. Even for that scenario there is room for improvement, but there are some options:

### GenericCollectionLayer
See the 'Demo |  Add Pins'

### WriteableLayer
Use the WriteableLayer, or your own implementation of it. It is used in the editing samples.

### Assign MemoryLayer.Features
Another way is to use a MemoryLayer and assign the MemoryLayer.Features with your data collection from which you can add and remove. Make sure it is a thread save collection if it gets mutated.

### Custom IProvider
Another way is to create your own provider:

```csharp
    public class MyProvider : IProvider
    {
        public string? CRS { get; set; } = "EPSG:4326"; // If they lat/lon coordinates

        public MRect? GetExtent()
        {
            // Return the extent of the entire data set. You may want to cache it, but update it when you modify the data.
        }

        public Task<IEnumerable<IFeature>> GetFeaturesAsync(FetchInfo fetchInfo)
        {
            // Return the features from which ever source you like.
        }

        public Add(IFeature feature)
        {
           // Add in a thread save way
        }
   }
```

**Reply by gigfy2** [reply]
I was trying and continued to try the MemoryLayer approach.  I've got

        ...
        ...
        ...
        map.Info += (s, e) =>
        {
            var memLayer = map.Layers.FirstOrDefault(_ => _.Name == "LineStringLayer") as MemoryLayer;

            Coordinate[] c = new Coordinate[] { new Coordinate(-100, 30), new Coordinate(-90, 25), new Coordinate(-80, 30) };
            var lineString = new LineString(c);
            var geometryFeature = new GeometryFeature { Geometry = lineString };

            if (memLayer != null)
            {
                ((List<GeometryFeature>)memLayer.Features).Add(geometryFeature);
            }

            memLayer?.DataHasChanged();
        };

        // DRAWING LINESTRING
        var lineStringLayer = CreateLineStringLayer(CreateLineStringStyle());
        map.Layers.Add(lineStringLayer);

        return Task.FromResult(map);
    }

    public static ILayer CreateLineStringLayer(IStyle? style = null)
    {
        Coordinate[] coordinates = new Coordinate[] { new Coordinate(-100, 40), new Coordinate(-90, 45), new Coordinate(-80, 40) };
        var lineString = new LineString(coordinates);

        lineString = new LineString(lineString.Coordinates.Select(v => SphericalMercator.FromLonLat(v.X, v.Y).ToCoordinate()).ToArray());

        return new MemoryLayer
        {
            Features = new[] { new GeometryFeature { Geometry = lineString } },
            Name = "LineStringLayer",
            Style = style

        };
    }

    public static IStyle CreateLineStringStyle()
    {
        return new VectorStyle
        {
            Fill = null,
            Outline = null,
            Line = { Color = Color.FromString("YellowGreen"), Width = 4 }
        };
    }
    
but, the line where I add the GeometryFeature on a click event

        ((List<GeometryFeature>)memLayer.Features).Add(geometryFeature);
        
I can't figure out why no cast is working.  IFeature, GeometryFeature, ... I get the can't cast to a list.  But, if it's not a list, I can't access the .Add().  So, I may have just hit a wall with frustration at this point, but IS this possible, what I'm trying to do?  Any idea what I need to cast to?

**Reply by pauldendulk** [reply]
Try this: `  ((List<GeometryFeature>)(memLayer.Features)).Add(geometryFeature);`

**Reply by gigfy2** [reply]
Yeah, that's one I've tried.  Here are those lines with the exception output.  It just doesn't want me to cast to a list for some reason.

                // System.InvalidCastException: 'Unable to cast object of type 'Mapsui.Nts.GeometryFeature[]' to type 'System.Collections.Generic.List`1[Mapsui.Nts.GeometryFeature]'.'
                ((List<GeometryFeature>)memLayer.Features).Add(geometryFeature);

                // System.InvalidCastException: 'Unable to cast object of type 'Mapsui.Nts.GeometryFeature[]' to type 'System.Collections.Generic.List`1[Mapsui.IFeature]'.'
                ((List<IFeature>)memLayer.Features).Add(geometryFeature);
                
I also tried it with the altered parenthesis, and got the same error

**Reply by gigfy2** [reply]
I finally got it.  The problem was that when creating the memory layer, I was setting Feature to a GeometryFeature.  I changed Feature to be a list of IFeature's and it works fine now.  Ah, so simple, but took forever to figure out.  Here's the code again, working, the relevant parts 

        map.Info += (s, e) =>
        {
            var memLayer = map.Layers.FirstOrDefault(_ => _.Name == "LineStringLayer") as MemoryLayer;

            if(e.MapInfo == null) { return; }

            var tappedPosition = SphericalMercator.ToLonLat(e.MapInfo.WorldPosition.X, e.MapInfo.WorldPosition.Y);

            Coordinate[] c = new Coordinate[] { new Coordinate(tappedPosition.lon, tappedPosition.lat), new Coordinate(-90, 25), new Coordinate(-80, 30) };
            var lineString = new LineString(c);
            lineString = new LineString(lineString.Coordinates.Select(v => SphericalMercator.FromLonLat(v.X, v.Y).ToCoordinate()).ToArray());

            var geometryFeature = new GeometryFeature { Geometry = lineString };

            if (memLayer != null)
            {
                ((List<IFeature>)memLayer.Features).Add(geometryFeature);
            }

            memLayer?.DataHasChanged();
        };

        // create the linestring layer and add it to the map
        var lineStringLayer = CreateLineStringLayer(CreateLineStringStyle());
        map.Layers.Add(lineStringLayer);

        return Task.FromResult(map);
    }

    public static ILayer CreateLineStringLayer(IStyle? style = null)
    {
        return new MemoryLayer
        {
            Features = new List<IFeature>(),
            Name = "LineStringLayer",
            Style = style
        };
    }
